File: src/scrapers/craigslist.js
const puppeteer = require("puppeteer-extra");
const StealthPlugin = require("puppeteer-extra-plugin-stealth");
puppeteer.use(StealthPlugin());

async function scrapeCraigslist({ maxRent, bedrooms }) {
  const browser = await puppeteer.launch({ headless: "new", args: ["--no-sandbox"] });
  const page = await browser.newPage();
  const listings = [];
  const brMap = { 1: "min_bedrooms=1&max_bedrooms=1", 2: "min_bedrooms=2&max_bedrooms=2", 3: "min_bedrooms=3" };
  const brParam = brMap[bedrooms] || "";
  const url = `https://chicago.craigslist.org/search/apa?max_price=${maxRent}&${brParam}#search=1~list~0~0`;
  try {
    await page.goto(url, { waitUntil: "networkidle2", timeout: 30000 });
    await page.waitForSelector(".cl-search-result", { timeout: 10000 });
    const results = await page.evaluate(() => {
      return Array.from(document.querySelectorAll(".cl-search-result")).map(el => {
        // Craigslist puts the neighborhood in .location > .housing-location or a nested <span>
        const hood =
          el.querySelector(".location .housing-location")?.innerText?.trim() ||
          el.querySelector(".meta .housing-location")?.innerText?.trim() ||
          el.querySelector("[data-hook='hood']")?.innerText?.trim() ||
          el.querySelector(".location")?.innerText?.trim() ||
          "";
        // Strip surrounding parens if present: "(Lincoln Park)" → "Lincoln Park"
        const neighborhood = hood.replace(/^\(|\)$/g, "").trim();
        const location = neighborhood ? `${neighborhood}, Chicago, IL` : "Chicago, IL";
        return {
          title:
            el.querySelector("a.posting-title .label")?.innerText?.trim() ||
            el.querySelector("a.posting-title")?.innerText?.trim() ||
            el.querySelector(".title")?.innerText?.trim() ||
            "",
          price:     el.querySelector(".priceinfo")?.innerText?.trim() || "",
          location,
          url:       el.querySelector("a.posting-title")?.href || "",
          source:    "Craigslist",
          scrapedAt: new Date().toISOString(),
        };
      });
    });
    listings.push(...results);
  } catch (err) {
    console.error("Craigslist scrape error:", err.message);
  }
  await browser.close();
  return listings;
}

module.exports = { scrapeCraigslist };

File: src/scrapers/craigslist.py
from datetime import datetime, timezone
from playwright.async_api import async_playwright


async def scrape_craigslist(filters: dict) -> list[dict]:
    max_rent = filters.get("maxRent")
    bedrooms = filters.get("bedrooms")

    br_map = {
        1: "min_bedrooms=1&max_bedrooms=1",
        2: "min_bedrooms=2&max_bedrooms=2",
        3: "min_bedrooms=3",
    }
    br_param = br_map.get(bedrooms, "")
    url = f"https://chicago.craigslist.org/search/apa?max_price={max_rent}&{br_param}#search=1~list~0~0"

    listings = []

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True, args=["--no-sandbox"])
        page = await browser.new_page()

        try:
            await page.goto(url, wait_until="networkidle", timeout=30000)
            await page.wait_for_selector(".cl-search-result", timeout=10000)

            results = await page.evaluate("""() => {
                return Array.from(document.querySelectorAll('.cl-search-result')).map(el => {
                    const hood =
                        el.querySelector('.location .housing-location')?.innerText?.trim() ||
                        el.querySelector('.meta .housing-location')?.innerText?.trim() ||
                        el.querySelector("[data-hook='hood']")?.innerText?.trim() ||
                        el.querySelector('.location')?.innerText?.trim() ||
                        '';
                    const neighborhood = hood.replace(/^\\(|\\)$/g, '').trim();
                    const location = neighborhood ? `${neighborhood}, Chicago, IL` : 'Chicago, IL';

                    const housing = el.querySelector('.housing')?.innerText?.trim() || '';
                    const bedsMatch = housing.match(/(\\d+)\\s*br/i);
                    const bathsMatch = housing.match(/(\\d+(?:\\.\\d+)?)\\s*ba/i);

                    return {
                        title:
                            el.querySelector('a.posting-title .label')?.innerText?.trim() ||
                            el.querySelector('a.posting-title')?.innerText?.trim() ||
                            el.querySelector('.title')?.innerText?.trim() ||
                            '',
                        price:     el.querySelector('.priceinfo')?.innerText?.trim() || '',
                        location,
                        url:       el.querySelector('a.posting-title')?.href || '',
                        beds:      bedsMatch ? bedsMatch[1] : '',
                        baths:     bathsMatch ? bathsMatch[1] : '',
                        source:    'Craigslist',
                        scrapedAt: new Date().toISOString(),
                    };
                });
            }""")

            listings.extend(results)

        except Exception as e:
            print(f"Craigslist scrape error: {e}")

        await browser.close()

    return listings
